drivers/overlay/mount.go

```diff
@@ -77,12 +79,111 @@ func mountFromMain() {
 		fatal(err)
 	}
 
-	if err := os.Chdir(flag.Arg(0)); err != nil {
+	// Mount the arguments passed from the specified directory. Some of the
+	// paths mentioned in the values we pass to the kernel are relative to
+	// the specified directory.
+	homedir := flag.Arg(0)
+	if err := os.Chdir(homedir); err != nil {
 		fatal(err)
 	}
 
-	if err := unix.Mount(options.Device, options.Target, options.Type, uintptr(options.Flag), options.Label); err != nil {
-		fatal(err)
+	pageSize := unix.Getpagesize()
+	if len(options.Label) < pageSize {
+		if err := unix.Mount(options.Device, options.Target, options.Type, uintptr(options.Flag), options.Label); err != nil {
+			fatal(err)
+		}
+		os.Exit(0)
+	}
+
+	// Those arguments still took up too much space.  Open the diff
+	// directories and use their descriptor numbers as lowers, using
+	// /proc/self/fd as the current directory.
+
+	// Split out the various options, since we need to manipulate the
+	// paths, but we don't want to mess with other options.
+	var upper, work, lower, label, other string
+	for _, arg := range strings.Split(options.Label, ",") {
+		if strings.HasPrefix(arg, "upperdir=") {
+			upper = strings.TrimPrefix(arg, "upperdir=")
+		} else if strings.HasPrefix(arg, "workdir=") {
+			work = strings.TrimPrefix(arg, "workdir=")
+		} else if strings.HasPrefix(arg, "lowerdir=") {
+			lower = strings.TrimPrefix(arg, "lowerdir=")
+		} else if strings.HasPrefix(arg, "label=") {
+			label = arg
```

Comment:
It’s a bit confusing that some variables contain just the data, and some the data + the option — and that for some, it _changes_ during the lifetime of this function.

Maybe … [A]